app/recipes/tests/test_recipes_api.py

```diff
@@ -0,0 +1,259 @@
+from django.urls import reverse
+from django.test import TestCase
+from rest_framework import status
+from rest_framework.test import APIClient
+
+from core.models import Recipe, Ingredient
+
+# Recipes collection resource URL, i.e. /recipes/
+RECIPES_URL = reverse('recipes:recipe-list')
+
+
+def recipe_url(recipe_id):
+    """Return recipe resource URL, i.e. /recipes/{recipe_id}/"""
+    return reverse('recipes:recipe-detail', args=[recipe_id])
+
+
+def create_sample_recipe(name, ingredient_names, description=''):
+    recipe = Recipe.objects.create(
+        name=name,
+        description=description,
+    )
+    for ingredient_name in ingredient_names:
+        Ingredient.objects.create(
+            name=ingredient_name,
+            recipe=recipe
+        )
+    return recipe
+
+
+class RecipeApiTests(TestCase):
+    """Test suite for recipes API endpoints"""
+
+    def setUp(self):
+        self.client = APIClient()
+
+    def test_create_recipe_with_description(self):
+        """POST /recipes/ should create recipe and return new resource
+            representation, including ID"""
+        payload = {
+            'name': 'Lemon cheesecake',
+            'description': 'Cheesecake (feat. lemon)',
+            'ingredients': [
+                {'name': 'Lemon'},
+                {'name': 'Cheese'}
+            ]
+        }
+
+        response = self.client.post(RECIPES_URL, payload, format='json')
+
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(response.json(), {
+            'id': response.data['id'],
+            'name': 'Lemon cheesecake',
+            'description': 'Cheesecake (feat. lemon)',
+            'ingredients': [
+                {'name': 'Lemon'},
+                {'name': 'Cheese'}
+            ]
+        })
+        created_recipe = Recipe.objects.get(id=response.data['id'])
+        self.assertEqual(created_recipe.name, 'Lemon cheesecake')
```

Comment:
IMO it is fine to just assert the total number of records (e.g. `assert Recipe.objects.count() == 1`) since the tests are isolated from each other, and no need to assert the contents of the records, since they are the only records that exists the previous assertions of the API response are testing content. Even then asserting DB records in this test is testing 2 levels of abstraction (API and DB) which probably isn't so great. For an exercise it is fine.

Reply:
>Even then asserting DB records in this test is testing 2 levels of abstraction (API and DB) which probably isn't so great. For an exercise it is fine.

Yeah, I thought we'd probably do it differently in real code. I was quite surprised that the "unit test" examples in the Udemy course seem like an off hybrid of end-to-end and integration tests, with some things happening in an implementation-indepedent way, and others being adjusted "under the hood".